File: backend/app/api/routes/patients.py
import uuid

from fastapi import APIRouter, HTTPException, Query, status
from sqlalchemy import select

from app.api.deps import CurrentUser, DbSession
from app.models.note import ClinicalNote
from app.models.patient import Patient
from app.models.user import User
from app.schemas.clinical import PatientOut, PatientRowOut, PatientStatusIn
from app.services.access import can_reset_practice, require_app_role
from app.services.audit import log_event
from app.services.serializers import apply_patient_snapshot, patient_out

router = APIRouter(prefix="/patients", tags=["patients"])


def _visible(p: Patient, viewer: User, app_role: str) -> bool:
    if app_role == "student":
        return p.mode == "practice" or p.owner_id == viewer.id
    return app_role in ("instructor", "admin", "front_desk")


@router.get("", response_model=list[PatientRowOut])
async def list_patients(
    user: CurrentUser,
    db: DbSession,
    courseId: uuid.UUID = Query(...),
    role: str = Query(...),
):
    role = await require_app_role(db, user.id, role, courseId)

    patients = (
        await db.scalars(
            select(Patient).where(
                Patient.course_id == courseId,
                Patient.deleted_at.is_(None),
            )
        )
    ).all()

    owner_ids = [p.owner_id for p in patients if p.owner_id]
    owners = {}
    if owner_ids:
        for u in (await db.scalars(select(User).where(User.id.in_(owner_ids)))).all():
            owners[u.id] = u.full_name

    rows: list[PatientRowOut] = []
    for p in patients:
        if not _visible(p, user, role):
            continue
        author_id = user.id if role == "student" else p.owner_id
        note_q = select(ClinicalNote).where(
            ClinicalNote.patient_id == p.id,
            ClinicalNote.archived.is_(False),
            ClinicalNote.deleted_at.is_(None),
        )
        if author_id:
            note_q = note_q.where(ClinicalNote.author_id == author_id)
        notes = (await db.scalars(note_q.order_by(ClinicalNote.updated_at.desc()))).all()
        latest = notes[0] if notes else None
        rows.append(
            PatientRowOut(
                patient=patient_out(p, owners.get(p.owner_id) if p.owner_id else None),
                ownerName=owners.get(p.owner_id) if p.owner_id else None,
                latestNote=(
                    {"status": latest.status, "updatedAt": latest.updated_at.isoformat()}
                    if latest
                    else None
                ),
            )
        )
    return rows


@router.get("/{patient_id}", response_model=PatientOut)
async def get_patient(
    patient_id: uuid.UUID,
    user: CurrentUser,
    db: DbSession,
    role: str = Query(...),
):
    p = await db.get(Patient, patient_id)
    if p is None or p.deleted_at is not None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="This patient doesn't exist or was archived.")
    role = await require_app_role(db, user.id, role, p.course_id)
    if not _visible(p, user, role):
        await log_event(
            db,
            action="chart.view",
            entity_type="patient",
            actor_user_id=user.id,
            entity_id=str(patient_id),
            details={"result": "denied"},
        )
        await db.commit()
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="This case belongs to another student. Each assessment case is private to the student it was assigned to.",
        )

    owner_name = None
    if role != "student" and p.owner_id:
        owner = await db.get(User, p.owner_id)
        owner_name = owner.full_name if owner else None

    await log_event(
        db,
        action="chart.view",
        entity_type="patient",
        actor_user_id=user.id,
        entity_id=str(patient_id),
        details={"result": "ok"},
    )
    await db.commit()
    return patient_out(p, owner_name)


@router.patch("/{patient_id}/status", response_model=PatientOut)
async def update_status(
    patient_id: uuid.UUID,
    body: PatientStatusIn,
    user: CurrentUser,
    db: DbSession,
    role: str = Query(...),
):
    p = await db.get(Patient, patient_id)
    if p is None or p.deleted_at is not None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Patient not found.")
    role = await require_app_role(db, user.id, role, p.course_id)

    patch = {k: v for k, v in body.model_dump().items() if v is not None}
    only_encounter = set(patch.keys()) <= {"encounter"}
    allowed = role in ("instructor", "admin") or (only_encounter and role in ("student", "instructor", "admin", "front_desk"))
    if not allowed:
        await log_event(
            db,
            action="patient.update_status",
            entity_type="patient",
            actor_user_id=user.id,
            entity_id=str(patient_id),
            details={"result": "denied"},
        )
        await db.commit()
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Only instructors can change this status.")

    p.status = {**(p.status or {}), **patch}
    await log_event(
        db,
        action="patient.update_status",
        entity_type="patient",
        actor_user_id=user.id,
        entity_id=str(patient_id),
        details={"result": "ok", "patch": patch},
    )
    await db.commit()
    await db.refresh(p)
    return patient_out(p)


@router.post("/{patient_id}/reset", status_code=status.HTTP_204_NO_CONTENT)
async def reset_practice(patient_id: uuid.UUID, user: CurrentUser, db: DbSession):
    if not can_reset_practice(user):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Only instructors can reset practice patients.")
    p = await db.get(Patient, patient_id)
    if p is None or p.mode != "practice" or not p.practice_snapshot:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Only practice patients can be reset.")

    apply_patient_snapshot(p, p.practice_snapshot)
    notes = (
        await db.scalars(
            select(ClinicalNote).where(
                ClinicalNote.patient_id == patient_id,
                ClinicalNote.archived.is_(False),
            )
        )
    ).all()
    for n in notes:
        n.archived = True

    await log_event(
        db,
        action="patient.reset_practice",
        entity_type="patient",
        actor_user_id=user.id,
        entity_id=str(patient_id),
        details={"result": "ok", "archived_notes": len(notes)},
    )
    await db.commit()

File: frontend/src/api/patients.ts
import type { NoteStatus, Patient, PatientStatus, Role, User } from "../types";
import { can } from "../utils/permissions";
import { ApiError, USE_MOCK, apiFetch, clone, latency } from "./client";
import { db } from "./mockDb";
import { recordAudit } from "./audit";

export interface PatientRow {
  patient: Patient;
  ownerName?: string;
  latestNote?: { status: NoteStatus; updatedAt: string };
}

function visibleTo(p: Patient, viewer: User, role: Role): boolean {
  if (role === "student") return p.mode === "practice" || p.ownerId === viewer.id;
  return role === "instructor" || role === "admin" || role === "front_desk";
}

export async function listPatients(viewer: User, role: Role, courseId: string): Promise<PatientRow[]> {
  if (!USE_MOCK) {
    return apiFetch<PatientRow[]>(`/patients?courseId=${encodeURIComponent(courseId)}&role=${encodeURIComponent(role)}`);
  }
  await latency();
  return db.patients
    .filter((p) => p.courseId === courseId && visibleTo(p, viewer, role))
    .map((p) => {
      const authorId = role === "student" ? viewer.id : p.ownerId;
      const latest = db.notes
        .filter((n) => n.patientId === p.id && (!authorId || n.authorId === authorId))
        .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt))[0];
      return {
        patient: clone(p),
        ownerName: db.users.find((u) => u.id === p.ownerId)?.fullName,
        latestNote: latest && { status: latest.status, updatedAt: latest.updatedAt },
      };
    });
}

export async function getPatient(viewer: User, role: Role, id: string): Promise<Patient> {
  if (!USE_MOCK) {
    return apiFetch<Patient>(`/patients/${id}?role=${encodeURIComponent(role)}`);
  }
  await latency();
  const p = db.patients.find((x) => x.id === id);
  if (!p) throw new ApiError(404, "This patient doesn't exist or was archived.");
  if (!visibleTo(p, viewer, role)) {
    recordAudit(viewer, "chart.view", `patient/${id}`, "denied", "Assessment case belongs to another student.");
    throw new ApiError(403, "This case belongs to another student. Each assessment case is private to the student it was assigned to.");
  }
  recordAudit(viewer, "chart.view", `patient/${id}`);
  const ownerName = role !== "student" ? db.users.find((u) => u.id === p.ownerId)?.fullName : undefined;
  return { ...clone(p), ownerName };
}

export async function updatePatientStatus(
  viewer: User, role: Role, id: string, patch: Partial<PatientStatus>,
): Promise<Patient> {
  if (!USE_MOCK) {
    return apiFetch<Patient>(`/patients/${id}/status?role=${encodeURIComponent(role)}`, {
      method: "PATCH",
      body: patch,
    });
  }
  await latency(150);
  const p = db.patients.find((x) => x.id === id);
  if (!p) throw new ApiError(404, "Patient not found.");
  const onlyEncounter = Object.keys(patch).every((k) => k === "encounter");
  const allowed = can(role, "patient:update_status") || (onlyEncounter && can(role, "encounter:advance"));
  if (!allowed) {
    recordAudit(viewer, "patient.update_status", `patient/${id}`, "denied");
    throw new ApiError(403, "Only instructors can change this status.");
  }
  p.status = { ...p.status, ...patch };
  recordAudit(viewer, "patient.update_status", `patient/${id}`, "ok", JSON.stringify(patch));
  return clone(p);
}

export async function resetPracticePatient(viewer: User, role: Role, id: string): Promise<void> {
  if (!USE_MOCK) {
    await apiFetch<void>(`/patients/${id}/reset`, { method: "POST" });
    return;
  }
  await latency(300);
  if (!can(role, "patient:reset_practice")) throw new ApiError(403, "Only instructors can reset practice patients.");
  const snapshot = db.practiceSnapshots.get(id);
  if (!snapshot) throw new ApiError(400, "Only practice patients can be reset.");
  const idx = db.patients.findIndex((p) => p.id === id);
  db.patients[idx] = structuredClone(snapshot);
  const [keep, archive] = [
    db.notes.filter((n) => n.patientId !== id),
    db.notes.filter((n) => n.patientId === id),
  ];
  db.notes.splice(0, db.notes.length, ...keep);
  db.archivedNotes.push(...archive);
  recordAudit(viewer, "patient.reset_practice", `patient/${id}`, "ok", `${archive.length} notes archived`);
}
